processing/src/main/java/io/druid/segment/StringDimensionQueryHelper.java

```diff
@@ -0,0 +1,443 @@
+  )
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    final int dimVal = keyBuffer.getInt();
+    if (dimVal != GROUP_BY_MISSING_VALUE) {
+      theEvent.put(outputName, selector.lookupName(dimVal));
+    }
+  }
+
+  @Override
+  public List<ByteBuffer> addDimValuesToGroupingKey(
+      Object selector, ByteBuffer key, Function<ByteBuffer, List<ByteBuffer>> updateValuesFn
+  )
+  {
+    List<ByteBuffer> unaggregatedBuffers = null;
+    final DimensionSelector dimSelector = (DimensionSelector) selector;
+    final IndexedInts row = dimSelector.getRow();
+    if (row == null || row.size() == 0) {
+      ByteBuffer newKey = key.duplicate();
+      newKey.putInt(GROUP_BY_MISSING_VALUE);
+      unaggregatedBuffers = updateValuesFn.apply(newKey);
+    } else {
+      for (Integer dimValue : row) {
+        ByteBuffer newKey = key.duplicate();
+        newKey.putInt(dimValue);
+        unaggregatedBuffers = updateValuesFn.apply(newKey);
+      }
+    }
+    return unaggregatedBuffers;
+  }
+
+  @Override
+  public Object getRowFromDimSelector(Object dimSelector) {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    IndexedInts values = selector == null ? EmptyIndexedInts.EMPTY_INDEXED_INTS : selector.getRow();
+    return values;
+  }
+
+  @Override
+  public int initializeGroupingKeyV2Dimension(
+      final Object valuesObj,
+      final ByteBuffer keyBuffer,
+      final int keyBufferPosition
+  )
+  {
+    IndexedInts values = (IndexedInts) valuesObj;
+    final int rowSize = values.size();
+    if (rowSize == 0) {
+      keyBuffer.putInt(keyBufferPosition, GROUP_BY_MISSING_VALUE);
+    } else {
+      keyBuffer.putInt(keyBufferPosition, values.get(0));
+    }
+    return rowSize;
+  }
+
+  @Override
+  public void addValueToGroupingKeyV2(
+      final Object values,
+      final int rowValueIdx,
+      final ByteBuffer keyBuffer,
+      final int keyBufferPosition
+  )
+  {
+    IndexedInts intValues = (IndexedInts) values;
+    keyBuffer.putInt(
+        keyBufferPosition,
+        intValues.get(rowValueIdx)
+    );
+  }
+
+  @Override
+  public void readValueFromGroupingKeyV2(QueryDimensionInfo dimInfo, Map<String, Object> resultMap, ByteBuffer key)
+  {
+    final int id = key.getInt(dimInfo.keyBufferPosition);
+
+    if (id >= 0) {
+      resultMap.put(
+          dimInfo.spec.getOutputName(),
+          ((DimensionSelector) dimInfo.selector).lookupName(id)
+      );
+    }
+  }
+
+  @Override
+  public Aggregator[][] getDimExtractionRowSelector(TopNParams params, TopNQuery query, Capabilities capabilities)
+  {
+    final BaseTopNAlgorithm.AggregatorArrayProvider provider = new BaseTopNAlgorithm.AggregatorArrayProvider(
+        (DimensionSelector) params.getDimSelector(),
+        query,
+        params.getCardinality(),
+        capabilities
+    );
+
+    // Unlike regular topN we cannot rely on ordering to optimize.
```

Comment:
The comment was copied over from the original location in DimExtractionTopNAlgorithm, it refers to why DimExtractionTopNAlgorithm would be used instead of other TopNAlgorithms (see getMapFn() in TopNQueryEngine)